src/test/java/hudson/plugins/ec2/EC2RetentionStrategyTest.java

```diff
@@ -215,18 +220,80 @@ public void testTerminateOfflineComputerIfNotConnecting() throws Exception {
+     *
+     * How does the test below work: for our "mock" EC2Computer, idle start time
+     * is always the time when the object has been created, and we cannot
+     * easily control idle start time in a test suite as the relevant method to override
+     * is declared final.
+     *
+     * We can achieve the same result where idle start time is < launch time
+     * by manipulating the node launch time variables. Since, as we said, idle
+     * start time is always now, what we end up doing is tricking the node
+     * into returning a launch time in the future. To do this we pass a negative
+     * value for minutes to {@link #computerWithUpTime(int, int, Boolean, Boolean)}.
+     *
+     * Now if we set retention time to something bigger than the computer uptime
+     * we can reproduce the issue, and prove its resolution.
+     *
+     * The time at which we perform the check must be < than computer uptime
+     * plus the retention interval.
+     */
+    @Test
+    public void testDoNotTerminateInstancesJustBooted() throws Exception {
+        logging.record(hudson.plugins.ec2.EC2RetentionStrategy.class, Level.FINE);
+        logging.capture(5);
+        final int COMPUTER_UPTIME_MINUTES=5;
+        final int RETENTION_MINUTES=5;
+        final int CHECK_TIME_MINUTES=COMPUTER_UPTIME_MINUTES+RETENTION_MINUTES-1;
+        final Instant checkTime = Instant.now().plus(Duration.ofMinutes(CHECK_TIME_MINUTES));
+        new EC2RetentionStrategy(
+            String.format("%d", RETENTION_MINUTES),
+            Clock.fixed(checkTime.plusSeconds(1), zoneId),
+            checkTime.toEpochMilli()
+        ).check(computerWithUpTime(-COMPUTER_UPTIME_MINUTES, 0, null, false));
+        assertThat("The computer is terminated, but should not be", idleTimeoutCalled.get(), equalTo(false));
+        assertThat(logging.getMessages(), hasItem(containsString("offline but not connecting, will check if it should be terminated because of the idle time configured")));
+    }
+
+    /**
+     * Ensure that we terminate instances that stay unconnected, as soon as
+     * termination time expires.
+     */
+    @Test
+    public void testCleanupUnconnectedInstanceAfterTerminationTime() throws Exception {
+        logging.record(hudson.plugins.ec2.EC2RetentionStrategy.class, Level.FINE);
+        logging.capture(5);
+        final int COMPUTER_UPTIME_MINUTES=5;
+        final int RETENTION_MINUTES=5;
+        final int CHECK_TIME_MINUTES=COMPUTER_UPTIME_MINUTES+RETENTION_MINUTES+1;
+        final Instant checkTime = Instant.now().plus(Duration.ofMinutes(CHECK_TIME_MINUTES));
+        new EC2RetentionStrategy(
+            String.format("%d", RETENTION_MINUTES),
+            Clock.fixed(checkTime.plusSeconds(1), zoneId),
+            checkTime.toEpochMilli()
+        ).check(computerWithUpTime(-COMPUTER_UPTIME_MINUTES, 0, null, false));
+        assertThat("The computer is terminated, but should not be", idleTimeoutCalled.get(), equalTo(true));
```

Comment:
Computer should be terminated in this case yes?

Reply:
thanks, fixed :-)